### ISSUE TITLE
Path challenge sent to wrong path after migration

### ISSUE THREAD

**Issue by kenmcmil** [private-octopus/picoquic#446]
The following log shows a case where a migration occurs from port 5047 to port 5048. However, instead of sending path challenges to both ports, the server sends two path challenges to port 5047 and continues to try to validate this path without ever validating the path to the new port, eventually timing out and closing the connection.

Also, notice that the server does not seem to recognize the PATH_RESPONSE to the first of the two challenges. It seems to indicate that the second PATH_CHALLENGE overwrote the first, which was possibly intended for the new path.  

```
Starting PicoQUIC server on port 4443, server name = ::, just_once = 0, hrr= 0
000000000000003d: Receiving 1232 bytes from 127.0.0.1:5047 at T=0.000000 (57d7e36c87452)
000000000000003d: Receiving packet type: 2 (initial), S0, Version ff000010,
000000000000003d:     <000000000000003d>, <000000000000003c>, Seq: 0, pl: 1207
000000000000003d:     Token length: 0
000000000000003d:     Decrypted 1187 bytes
000000000000003d:     Crypto HS frame, offset 0, length 255: 010000fb0303284c...
000000000000003d:     padding, 928 bytes

3d: 0.002320 : Connection established, state = 11, from length: 16
3d: Client address: 127.0.0.1, port 46867
000000000000003d: SNI not received.
000000000000003d: Received ALPN: hq-16
000000000000003d: Received transport parameter TLS extension (44 bytes):
3d: Proposed version: ff000010
3d:     Extension list (38 bytes):
3d:         Extension type: 0, length 4 (0x0000 / 0x0004), 00002000
3d:         Extension type: 1, length 4 (0x0001 / 0x0004), 00004000
3d:         Extension type: 3, length 2 (0x0003 / 0x0002), 003c
3d:         Extension type: 10, length 4 (0x000a / 0x0004), 00002000
3d:         Extension type: 11, length 4 (0x000b / 0x0004), 00002000
000000000000003d: Sending packet type: 2 (initial), S0, Version ff000010,
000000000000003d:     <000000000000003c>, <e3e4b9d1983d775e>, Seq: 0, pl: 184
000000000000003d:     Token length: 0
000000000000003d:     Prepared 164 bytes
000000000000003d:     ACK (nb=0), 0
000000000000003d:     Crypto HS frame, offset 0, length 155: 0200009703039c0b...

000000000000003d: Sending packet type: 4 (handshake), S0, Version ff000010,
000000000000003d:     <000000000000003c>, <e3e4b9d1983d775e>, Seq: 0, pl: 1019
000000000000003d:     Prepared 999 bytes
000000000000003d:     Crypto HS frame, offset 0, length 995: 0800007a00780010...

000000000000003d: Sending 1252 bytes to 127.0.0.1:5047 at T=0.000000 (57d7e36c87452)
3d: Connection state = 6
000000000000003d: Sending packet type: 4 (handshake), S0, Version ff000010,
000000000000003d:     <000000000000003c>, <e3e4b9d1983d775e>, Seq: 1, pl: 300
000000000000003d:     Prepared 280 bytes
000000000000003d:     Crypto HS frame, offset 995, length 275: d46d6533944770a0...

000000000000003d: Sending packet type: 6 (1rtt protected), S0,
000000000000003d:     <000000000000003c>, Seq: 0 (0), Phi: 0,
000000000000003d:     Prepared 216 bytes
000000000000003d:     NEW CONNECTION ID[1]: 0xdd7a5e6bef241629, bbf21ddd0d76d462c866d051ed0a9ff2
000000000000003d:     NEW CONNECTION ID[2]: 0x271ab3d559677bbc, 6de3d26624657f1619de06304c3c38ce
000000000000003d:     NEW CONNECTION ID[3]: 0xe927197e2f5e0809, 5233634aa72967d1261c5b10a828e017
000000000000003d:     NEW CONNECTION ID[4]: 0x1d0c6a0d41d7b40f, 7c8912214b4029e4664a9e355ff394b4
000000000000003d:     NEW CONNECTION ID[5]: 0xa6b504f88c9c1663, 330736d8a661eb3ad74e49d610bbdd62
000000000000003d:     NEW CONNECTION ID[6]: 0xc592ad472fc47315, 37dd1cbafdf437d77ca7446340e6c0be
000000000000003d:     NEW CONNECTION ID[7]: 0x19f3ca91fceb64c7, e4d4465d80d33eefecb5616e87bb8e3c
000000000000003d:     NEW CONNECTION ID[8]: 0x7a748ae02769e684, 4c64fb53bb0ca2b424277259c60f9ff3

000000000000003d: Sending 569 bytes to 127.0.0.1:5047 at T=0.000000 (57d7e36c87452)
000000000000003d: Sending packet type: 6 (1rtt protected), S0,
000000000000003d:     <000000000000003c>, Seq: 1 (1), Phi: 0,
000000000000003d:     Prepared 1411 bytes
000000000000003d:     ping, 1 bytes
000000000000003d:     padding, 1410 bytes

000000000000003d: Sending 1440 bytes to 127.0.0.1:5047 at T=0.000000 (57d7e36c87452)
000000000000003d: Receiving 120 bytes from 127.0.0.1:5047 at T=0.177443 (57d7e36cb2975)
000000000000003d: Receiving packet type: 4 (handshake), S0, Version ff000010,
000000000000003d:     <e3e4b9d1983d775e>, <000000000000003c>, Seq: 0, pl: 96
000000000000003d:     Decrypted 76 bytes
000000000000003d:     Crypto HS frame, offset 0, length 52: 1400003044366b13...
000000000000003d:     ACK (nb=0), 1
000000000000003d:     padding, 16 bytes

000000000000003d: Sending packet type: 4 (handshake), S0, Version ff000010,
000000000000003d:     <000000000000003c>, <e3e4b9d1983d775e>, Seq: 2, pl: 25
000000000000003d:     Prepared 5 bytes
000000000000003d:     ACK (nb=0), 0

000000000000003d: Sending 49 bytes to 127.0.0.1:5047 at T=0.177443 (57d7e36cb2975)
000000000000003d: Receiving 47 bytes from 127.0.0.1:5047 at T=0.310913 (57d7e36cd32d3)
000000000000003d: Receiving packet type: 6 (1rtt protected), S0,
000000000000003d:     <e3e4b9d1983d775e>, Seq: 0 (0), Phi: 0,
000000000000003d:     Decrypted 18 bytes
000000000000003d:     MAX DATA: 0x28.
000000000000003d:     padding, 16 bytes

000000000000003d: Sending packet type: 6 (1rtt protected), S0,
000000000000003d:     <000000000000003c>, Seq: 2 (2), Phi: 0,
000000000000003d:     Prepared 24 bytes
000000000000003d:     ACK (nb=0), 0
000000000000003d:     padding, 19 bytes

000000000000003d: Sending 53 bytes to 127.0.0.1:5047 at T=0.310913 (57d7e36cd32d3)
000000000000003d: Receiving 49 bytes from 127.0.0.1:5047 at T=0.398215 (57d7e36ce87d9)
000000000000003d: Receiving packet type: 6 (1rtt protected), S1,
000000000000003d:     <e3e4b9d1983d775e>, Seq: 2 (2), Phi: 0,
000000000000003d:     Decrypted 20 bytes
000000000000003d:     STOP SENDING 4 (0x00000004), Error 0x12ad.
000000000000003d:     padding, 16 bytes

3d: 0.398237 : Server CB, Stream: 4, 0 bytes, fin=3 (stop sending)
3d: Server CB, Stop Sending Stream: 4, resetting the local stream.
000000000000003d: Sending packet type: 6 (1rtt protected), S1,
000000000000003d:     <000000000000003c>, Seq: 3 (3), Phi: 0,
000000000000003d:     Prepared 24 bytes
000000000000003d:     ACK (nb=1), 2, 0
000000000000003d:     RESET STREAM 4, Error 0x00000000, Offset 0x0.
000000000000003d:     padding, 12 bytes

000000000000003d: Sending 53 bytes to 127.0.0.1:5047 at T=0.398215 (57d7e36ce87d9)
000000000000003d: Sending packet type: 6 (1rtt protected), S1,
000000000000003d:     <000000000000003c>, Seq: 4 (4), Phi: 0,
000000000000003d:     Prepared 226 bytes
000000000000003d:     NEW CONNECTION ID[1]: 0xdd7a5e6bef241629, bbf21ddd0d76d462c866d051ed0a9ff2
000000000000003d:     NEW CONNECTION ID[2]: 0x271ab3d559677bbc, 6de3d26624657f1619de06304c3c38ce
000000000000003d:     NEW CONNECTION ID[3]: 0xe927197e2f5e0809, 5233634aa72967d1261c5b10a828e017
000000000000003d:     NEW CONNECTION ID[4]: 0x1d0c6a0d41d7b40f, 7c8912214b4029e4664a9e355ff394b4
000000000000003d:     NEW CONNECTION ID[5]: 0xa6b504f88c9c1663, 330736d8a661eb3ad74e49d610bbdd62
000000000000003d:     NEW CONNECTION ID[6]: 0xc592ad472fc47315, 37dd1cbafdf437d77ca7446340e6c0be
000000000000003d:     NEW CONNECTION ID[7]: 0x19f3ca91fceb64c7, e4d4465d80d33eefecb5616e87bb8e3c
000000000000003d:     NEW CONNECTION ID[8]: 0x7a748ae02769e684, 4c64fb53bb0ca2b424277259c60f9ff3
000000000000003d:     ACK (nb=1), 2, 0

000000000000003d: Sending 255 bytes to 127.0.0.1:5047 at T=0.532540 (57d7e36d0948e)
000000000000003d: Receiving 69 bytes from 127.0.0.1:5048 at T=0.668061 (57d7e36d2a5ef)
000000000000003d: Receiving packet type: 6 (1rtt protected), S0,
000000000000003d:     <e3e4b9d1983d775e>, Seq: 3 (3), Phi: 0,
000000000000003d:     Decrypted 40 bytes
000000000000003d:     Stream 16, offset 0, length 16, fin = 1: 474554202f696e64...
000000000000003d:     STOP SENDING 4 (0x00000004), Error 0xdb3.
000000000000003d:     padding, 16 bytes

3d: 0.668086 : Server CB, Stream: 16, 16 bytes, fin=1 (stream fin)
3d: Server CB, Stream: 16, Processing command: GET /index.html
000000000000003d: Sending packet type: 6 (1rtt protected), S0,
000000000000003d:     <000000000000003c>, Seq: 5 (5), Phi: 0,
000000000000003d:     Prepared 1203 bytes
000000000000003d:     path_challenge: a955e742bbf828cb
000000000000003d:     padding, 1194 bytes

000000000000003d: Sending 1232 bytes to 127.0.0.1:5047 at T=0.668061 (57d7e36d2a5ef)
000000000000003d: Sending packet type: 6 (1rtt protected), S0,
000000000000003d:     <000000000000003c>, Seq: 6 (6), Phi: 0,
000000000000003d:     Prepared 16 bytes
000000000000003d:     path_challenge: c7d411017f6edfbb
000000000000003d:     ACK (nb=1), 2-3, 0

000000000000003d: Sending 45 bytes to 127.0.0.1:5047 at T=0.668061 (57d7e36d2a5ef)
000000000000003d: Receiving 1232 bytes from 127.0.0.1:5047 at T=0.848270 (57d7e36d565e0)
000000000000003d: Receiving packet type: 2 (initial), S0, Version ff000010,
000000000000003d:     <e3e4b9d1983d775e>, <000000000000003c>, Seq: 12, pl: 1207
000000000000003d:     Token length: 0
000000000000003d:     Decrypted 1187 bytes
000000000000003d:     ACK (nb=0), 0
000000000000003d:     padding, 1182 bytes

000000000000003d: Sending packet type: 6 (1rtt protected), S0,
000000000000003d:     <000000000000003c>, Seq: 7 (7), Phi: 0,
000000000000003d:     Prepared 11 bytes
000000000000003d:     ping, 1 bytes
000000000000003d:     ACK (nb=1), 2-3, 0

000000000000003d: Sending 40 bytes to 127.0.0.1:5047 at T=1.000197 (57d7e36d7b757)
000000000000003d: Receiving 98 bytes from 127.0.0.1:5047 at T=1.137431 (57d7e36d9cf69)
000000000000003d: Receiving packet type: 6 (1rtt protected), S0,
000000000000003d:     <e3e4b9d1983d775e>, Seq: 19 (19), Phi: 0,
000000000000003d:     Decrypted 69 bytes
000000000000003d:     Stream 4, offset 0, length 16, fin = 1: 474554202f696e64...
000000000000003d:     STOP SENDING 4 (0x00000004), Error 0x5df5.
000000000000003d:     path_response: a955e742bbf828cb
000000000000003d:     Stream 20, offset 0, length 16, fin = 1: 474554202f696e64...
000000000000003d:     padding, 16 bytes

3d: 1.137459 : Server CB, Stream: 4, 16 bytes, fin=1 (stream fin)
3d: Server CB, Stream: 4, RESET, too long or after FIN
3d: 1.137463 : Server CB, Stream: 20, 16 bytes, fin=1 (stream fin)
3d: Server CB, Stream: 20, Processing command: GET /index.html
000000000000003d: Sending packet type: 6 (1rtt protected), S0,
000000000000003d:     <000000000000003c>, Seq: 8 (8), Phi: 0,
000000000000003d:     Prepared 19 bytes
000000000000003d:     path_challenge: c7d411017f6edfbb
000000000000003d:     ACK (nb=2), 13, 2-3, 0

000000000000003d: Sending 48 bytes to 127.0.0.1:5047 at T=1.200507 (57d7e36dac5cd)
000000000000003d: Sending packet type: 6 (1rtt protected), S0,
000000000000003d:     <000000000000003c>, Seq: 9 (9), Phi: 0,
000000000000003d:     Prepared 21 bytes
000000000000003d:     path_challenge: c7d411017f6edfbb
000000000000003d:     ACK (nb=2), 13, 2-3, 0

000000000000003d: Sending 50 bytes to 127.0.0.1:5047 at T=1.733422 (57d7e36e2e780)
000000000000003d: Sending packet type: 6 (1rtt protected), S0,
000000000000003d:     <000000000000003c>, Seq: 10 (10), Phi: 0,
000000000000003d:     Prepared 21 bytes
000000000000003d:     path_challenge: c7d411017f6edfbb
000000000000003d:     ACK (nb=2), 13, 2-3, 0

000000000000003d: Sending 50 bytes to 127.0.0.1:5047 at T=2.266360 (57d7e36eb094a)
000000000000003d: Sending packet type: 6 (1rtt protected), S0,
000000000000003d:     <000000000000003c>, Seq: 11 (11), Phi: 0,
000000000000003d:     Prepared 17 bytes
000000000000003d:     RESET STREAM 4, Error 0x00000000, Offset 0x0.
000000000000003d:     ACK (nb=2), 13, 2-3, 0

000000000000003d: Sending 46 bytes to 127.0.0.1:5047 at T=2.398320 (57d7e36ed0cc2)
3d: 2.799205 : Server CB, Stream: 0, 0 bytes, fin=5 (connection close)
000000000000003d: Sending packet type: 6 (1rtt protected), S0,
000000000000003d:     <000000000000003c>, Seq: 12 (12), Phi: 0,
000000000000003d:     Prepared 21 bytes
000000000000003d:     path_challenge: c7d411017f6edfbb
000000000000003d:     ACK (nb=2), 13, 2-3, 0

000000000000003d: Sending 50 bytes to 127.0.0.1:5047 at T=2.799162 (57d7e36f32a8c)
3d: Connection state = 19
3d: 2.799280 : Closed. Retrans= 3, spurious= 0, max sp gap = 0, max sp delay = 0
e3e4b9d1983d775e: Receiving 273 bytes from 127.0.0.1:5047 at T=0.000000 (57d7e370d95e5)
e3e4b9d1983d775e: Receiving packet type: 6 (1rtt protected), S0,
e3e4b9d1983d775e:     <e3e4b9d1983d775e>, Seq: 0 (0), Phi: 0,
e3e4b9d1983d775e:     Decrypted 264 bytes
e3e4b9d1983d775e:     Unknown frame, type: 1947595446826941819

e3e4b9d1983d775e: Unexpected connection ID, sending stateless reset.
e3e4b9d1983d775e: Sending 81 bytes to 127.0.0.1:5047 at T=0.000000 (57d7e370d9606)


```

**Comment by huitema** [comment]
Yes, this is bad. Part of the draft-17 work is to make sure that probe responses are only accepted if they come from the right addresses. That means this bug will be fixed just after we finish the migration to draft-17.

**Comment by huitema** [comment]
@kenmcmil I implemented the required address checks in PR #449. I do not observe that the code is sending challenges to the wrong addresses, but the lack of checks caused an interaction when the NAT was rewriting some of the packet addresses, and the code could sometimes accept challenge responses from the wrong address.

Can you describe the test setup that you used to find this problem?

**Comment by huitema** [comment]
Verified that the path challenges are properly sent to two addresses:
```
ffe113a5348d6026: Sending packet type: 6 (1rtt protected), S0,
ffe113a5348d6026:     <f304a2bfdab356bf>, Seq: 20 (20), Phi: 0,
ffe113a5348d6026:     Prepared 1203 bytes
ffe113a5348d6026:     path_challenge: d4f98831acbf480a
ffe113a5348d6026:     padding, 1194 bytes
ffe113a5348d6026: Sending 1232 bytes to 172.56.42.39:58028 at T=0.599439 (57e1da72ccf1b)
ffe113a5348d6026: Sending packet type: 4 (handshake), S0, Version 50435131,
ffe113a5348d6026:     <f304a2bfdab356bf>, <6e386e1f13f21a15>, Seq: 2, pl: 25
ffe113a5348d6026:     Prepared 5 bytes
ffe113a5348d6026:     ACK (nb=0), 0-1
ffe113a5348d6026: Sending packet type: 6 (1rtt protected), S0,
ffe113a5348d6026:     <f304a2bfdab356bf>, Seq: 21 (21), Phi: 0,
ffe113a5348d6026:     Prepared 16 bytes
ffe113a5348d6026:     path_challenge: 6283167b777de883
ffe113a5348d6026:     ACK (nb=1), 3, 0-1
ffe113a5348d6026: Sending 94 bytes to 172.56.42.39:47076 at T=0.599439 (57e1da72ccf1b)
```
